tools/x-repo-types/typeAnalyzer/typeAnalyzer.go

```diff
@@ -0,0 +1,575 @@
+
+	valChild := TypeNode{t.Depth + 1, valueType, valueType.Kind(), nil, nil}
+	path = valChild.build(path)
+	t.appendChild("<map val>", "", valChild)
+	return path
+}
+
+func (t *TypeNode) buildPtrChild(path TypePath) TypePath {
+	tt := t.Type.Elem()
+	child := TypeNode{t.Depth + 1, tt, tt.Kind(), nil, nil}
+	path = child.build(path)
+	t.appendChild("<ptr elt>", "", child)
+	return path
+}
+
+// Visit traverses the Target tree and applies any actions provided at each node.
+func (tgt *Target) Visit(actions ...func(Target)) {
+	if len(actions) > 0 {
+		for _, action := range actions {
+			action(*tgt)
+		}
+		for _, target := range tgt.TypeNode.Targets() {
+			target.Visit(actions...)
+		}
+	}
+}
+
+// StructDiff compares two structs by building their type tree and then
+// calling targetTreeDiff on the trees.
+func StructDiff(x, y interface{}, exclusions map[string]bool) (TypeNode, TypeNode, *Diff, error) {
+	xRoot, yRoot := MakeType(x), MakeType(y)
+	xRoot.Build()
+	yRoot.Build()
+
+	diff, err := targetTreeDiff(Target{TypeNode: xRoot}, Target{TypeNode: yRoot}, exclusions)
+	return xRoot, yRoot, diff, err
+}
+
+// targetTreeDiff recursively computes a diff between two Target's x and y considering only data
+// that impacts serialization, but ignoring any assumption on key ordering.
+func targetTreeDiff(x, y Target, exclusions map[string]bool) (*Diff, error) {
+	xtype, ytype := x.TypeNode, y.TypeNode
+	if xtype.Depth != ytype.Depth {
+		return nil, fmt.Errorf("cannot compare types at different depth")
+	}
+	// if we got here it must be the case that either depth == 0 or
+	// the edges of x and y serialize the same way.
+
+	// First check that the native type for x and y are the same.
+	if xtype.Kind != ytype.Kind {
+		return &Diff{
+			Xdiff: []Target{x},
+			Ydiff: []Target{y},
+		}, nil
+	}
+
+	// So look at the children.
+	// If any children differ report back the diff.
+	xTgts, yTgts := xtype.Targets(), ytype.Targets()
+	xSerials, ySerials := make(map[string]Target), make(map[string]Target)
+	for _, tgt := range xTgts {
+		xSerials[tgt.ChildName.SerializationInfo()] = tgt
+	}
+	for _, tgt := range yTgts {
+		ySerials[tgt.ChildName.SerializationInfo()] = tgt
+	}
+	xDiff, yDiff := []Target{}, []Target{}
+	for k, v := range xSerials {
+		if _, ok := ySerials[k]; !ok {
+			xDiff = append(xDiff, v)
+		}
+	}
+	for k, v := range ySerials {
+		if _, ok := xSerials[k]; !ok {
+			yDiff = append(yDiff, v)
+		}
+	}
+	if len(xDiff) != 0 || len(yDiff) != 0 {
+		return &Diff{
+			Xdiff: xDiff,
+			Ydiff: yDiff,
+		}, nil
+	}
+
+	// Otherwise, call the children recursively. If any of them report
+	// a diff, modify the diff's CommonPath to include the current edge and return it.
+	for k, xChild := range xSerials {
+		if exclusions[xChild.TypeNode.String()] {
+			continue
+		}
+		yChild := ySerials[k]
+		diff, err := targetTreeDiff(xChild, yChild, exclusions)
+		if err != nil {
+			return nil, err
+
+		}
+		if diff != nil {
```

Comment:
I see what you mean by terminating at the first diff. This seems fine, it can always be extended if we decide a more complete report is needed for some reason.